conn_pool.go

```diff
@@ -248,7 +282,36 @@ func (p *ConnPool) createConnection() (*Conn, error) {
 	if err != nil {
 		return nil, err
 	}
+	return p.afterConnectionCreated(c)
+}
+
+// createConnectionUnlocked Removes the current lock, creates a new connection, and
+// then locks it back.
+// Here is the point: lets say our pool dialer's OpenTimeout is set to 3 seconds.
+// And we have a pool with 20 connections in it, and we try to acquire them all at
+// startup.
+// If it happens that the remote server is not accessible, then the first connection
+// in the pool blocks all the others for 3 secs, before it gets the timeout. Then
+// connection #2 holds the lock and locks everything for the next 3 secs until it
+// gets OpenTimeout err, etc. And the very last 20th connection will fail only after
+// 3 * 20 = 60 secs.
+// To avoid this we put Connect(p.config) outside of the lock (it is thread safe)
+// what would allow us to make all the 20 connection in parallel (more or less).
+func (p *ConnPool) createConnectionUnlocked() (*Conn, error) {
+	p.cond.L.Unlock()
+	c, err := Connect(p.config)
+	p.cond.L.Lock()
+
+	if err != nil {
+		return nil, err
+	}
+	return p.afterConnectionCreated(c)
```

Comment:
should `createConnectionUnlocked` check `p.resetCount` before the lock is released, and after the lock is reacquired?
If the `resetCount` has changed meanwhile, this connection should be discarded?